added selinux support

ensc opened the pull request:
Commit adds SELinux support to dropbear by:

- adding a new '--enable-selinux' option to configure; by default, it
  is disabled.  This option defines an ENABLE_SELINUX preprocessor
  macro.

- mapping the unix username to the SELinux user which is stored in a
  new 'user_sid' attribute in the AuthState object

- relabeling the controlling pty

- setting the context for the next execve() call to the user_sid


Operations above will not be done when SELinux is disabled.  Failures will
generate LOG_ERR messages and in enforcing SELinux mode, dropbear_exit()
will be called.


Signed-off-by: Enrico Scholz <enrico.scholz@ensc.de>

ensc committed:
added selinux support

Commit adds SELinux support to dropbear by:

- adding a new '--enable-selinux' option to configure; by default, it
  is disabled.  This option defines an ENABLE_SELINUX preprocessor
  macro.

- mapping the unix username to the SELinux user which is stored in a
  new 'user_sid' attribute in the AuthState object

- relabeling the controlling pty

- setting the context for the next execve() call to the user_sid

Operations above will not be done when SELinux is disabled.  Failures will
generate LOG_ERR messages and in enforcing SELinux mode, dropbear_exit()
will be called.

Signed-off-by: Enrico Scholz <enrico.scholz@ensc.de>

mkj reviewed (changes requested):
Thanks. I've noted a few changes.
Out of interest what platform are you using?

ensc commented on Makefile.in:
```diff
@@ -24,6 +24,11 @@ CFLAGS+=-I$(srcdir)/libtomcrypt/src/headers/
 
+ifeq (@ENABLE_SELINUX@, 1)
+LIBS+=-lselinux
+CFLAGS+=-DENABLE_SELINUX
```
ok; I replaced this with some more autoconf/autoheader integration:

- the `DROPBEAR_ENABLE_SELINUX` macro is now defined  automatically in `config.h`
- there was added a `LIBSELINUX_LIBS` makefile variable (which can be overridden at `./configure` time) which is applied to server build only

ensc commented on auth.h:
```diff
@@ -127,6 +127,10 @@ struct AuthState {
 	struct PubKeyOptions* pubkey_options;
 #endif
+
+#ifdef ENABLE_SELINUX
```
I use now everywhere

```
#ifdef DROPBEAR_ENABLE_SELINUX
```

ensc commented on common-session.c:
```diff
@@ -615,6 +663,9 @@ void fill_passwd(const char* username) {
 	ses.authstate.pw_dir = m_strdup(pw->pw_dir);
 	ses.authstate.pw_shell = m_strdup(pw->pw_shell);
+
+	initselinux(ses.authstate.pw_name);
```
ok; I redesigned this so that SELinux related code is executed only in `svr-*.c`.

ensc commented on svr-chansession.c:
```diff
@@ -556,6 +560,48 @@ static void get_termmodes(const struct ChanSess *chansess) {
+		goto out;
+	}
+
+	class = string_to_security_class("chr_file");
```
I added a check

ensc commented:
I think, I addressed all the issues.

I am using it in an OpenEmbedded based project:

- see https://gitlab.com/ensc-groups/bpi-router for toplevel project
- the SELinux related changes are in https://gitlab.com/ensc-groups/bpi-router/de.ensc.bpi-router/tree/master/meta-core/recipes-selinux/dropbear
- the dropbear SELinux policy (based upon recent 'minimum' policy) is in https://gitlab.com/ensc-groups/bpi-router/tools/selinux-policy/tree/master/policy; perhaps labeling `dropbear` files as `sshd_exec_t` will work too but I never tried it
- you will need something like https://gitlab.com/ensc-groups/bpi-router/tools/selinux-policy/blob/master/contexts/default_contexts

ensc committed:
Merge branch 'master' into selinux
